File: MicroBenchmarks/Ju/main_lapl_1d.jl

# initalize an array
function Init!(X,L)
    size=length(X)
    h=L/size
    for i=0:size-1
        if i>floor(size//8) && i<floor(size//2)+floor(size//8)
            X[i+1]=1.-2*(i-floor(size/8))*h/L;
        else
            X[i+1]=0.0
        end
    end
end



# proc[1-5]! functions make the same computation, with different programing
# styles.
function proc1!(In::Array{Float64,1},Out::Array{Float64,1},niter::Int64)
    const size=length(In)
    h2::Float64= (1./size)^2
    for it=1:niter
        Out[2:size-1]= h2*(In[1:size-2]- 2.0*In[2:size-1]+ In[3:size])
        In,Out=Out,In
    end 
end
function proc1!(In::Array{Float64,1},Out::Array{Float64,1},niter::Int64)
    const size=length(In)
    h2::Float64= (1./size)^2
    for it=1:niter
        Out[2:size-1]= h2*(In[1:size-2]- 2.0*In[2:size-1]+ In[3:size])
        In,Out=Out,In
    end 
end
function proc2!(In::Array{Float64,1},Out::Array{Float64,1},niter::Int64)
    const size=length(In)
    h2::Float64= (1./size)^2
    for it=1:niter
        @.  Out[2:size-1]= h2*(In[1:size-2]- 2.0*In[2:size-1]+ In[3:size])
        In,Out=Out,In
    end 
end
function proc3!(In::Array{Float64,1},Out::Array{Float64,1},niter::Int64)
    const size=length(In)
    h2::Float64= (1./size)^2
    for it=1:niter
        @simd for i=2:size-1
            Out[i]= h2*(In[i-1]- 2.0*In[i]+ In[i+1])
        end
        In,Out=Out,In
    end 
end
# measure as precisely as possible the execution time for a given
# p= proc[1-5]!
function test(p,In::Array{Float64,1},Out::Array{Float64,1},nit::Int)
   
    niter=nit
    Init!(In,1.)
    Init!(Out,1.)
 
    p(In,Out,niter)
    T=0.
    while true
        Init!(In,1.)
        Init!(Out,1.)
        t1 = time_ns()
        p(In,Out,niter)
        t = (time_ns() -t1)/niter
        if abs(t-T)/t<0.025
            break
    else
            T=t
            if niter> 150000
                error("too many iterations",niter)
            else
                niter*=2
            end
    end
    
end
    T,niter
end
DD=Dict("proc1!"=>"Vectorisation with braces         ",
        "proc2!"=>"Vectorisation with @. and braces  ",
        "proc3!"=>"Unrolled loop                     "
        )

fw=open("RunningOn"*gethostname()*"_lapl_1","w")

# computation starts here:
size=32
sizemax=10^5
const niter=2
while size<sizemax
    println("size: ",size)
    In=Array{Float64}(size)
    Out=Array{Float64}(size)
    tbest=10.^29
    best=0
    t=0.0
    for p= [proc1!,proc2!,proc3!]
        t,it=test(p,In,Out,niter)
        if t<tbest
            tbest=t
            best=p
        end
        t*=10.0^(-9)
        println(DD[string(p)]," : t= ",t," seconds ")
    end

    nflops=size*4
    flops=nflops/tbest
    println("\nbest: ",DD[string(best)])

    write(fw,string(size)," ",string(tbest*10.0^(-9)),"\n")
    
    println("nb. flops: ",nflops, ", Gflops/s (best): ",flops)
    println("-------")
    
    size*=2
    println()
end
close(fw)

File: MicroBenchmarks/Py/main_lapl_1d.py
import numpy as np
import time
import socket

def Init(X,L):
    size=X.size
    h=L/size
    for i in range(0,size):
        if i>size//8 and i<size//2+size//8:
            X[i]=1.-2*(i-size//8)*h/L;
        else:
            X[i]=0.0


def lapl1d_1(In,Out,niter):
    size=In.size
    h2= (1./size)**2
    for it in range(0,niter):
        Out[1:size-1]= h2*(In[0:size-2]- 2.0*In[1:size-1]+ In[2:size])
        In,Out=Out,In

def lapl1d_2(In,Out,niter):
    size=In.size
    h2= (1./size)**2
    for it in range(0,niter):
        for i in range(1,size-1):
            Out[i]=h2*(In[i-1]- 2.0*In[i]+ In[i+1])
        In,Out=Out,In  
      
def test(p,In,Out,nit):
   
    niter=nit
    Init(In,1.)
    Init(Out,1.)
 
    p(In,Out,niter)
    T=0.
    while True:
        Init(In,1.)
        Init(Out,1.)
        t1 = time.time()
        p(In,Out,niter)
        treal=time.time() -t1
        t = treal/niter
        if treal>0.0001 and abs(t-T)/t<0.025:
            break
        else:
            T=t
            niter*=2

    return T,niter

DD={"lapl1d_1":"Vectorized",
    "lapl1d_2":"Naïve     "}

f=open("RunningOn"+socket.gethostname()+"_lapl_1","w")

size=16
sizemax=100000
niter=10
parsef= lambda  f: str(f).split(" ")[1] #parse function name
while size<sizemax:
    print("size: ",size)
    In= np.empty(size)
    Out= np.empty(size)
    tbest=10.**20
    best=0
    t=0.0
    for p in  [lapl1d_1,lapl1d_2]:
        t,it=test(p,In,Out,niter)
        if t<tbest:
            tbest=t
            best=p
        print(DD[parsef(p)]," : t= ",t," seconds ")
    nflops= 4*(size-2)
    flops=nflops/tbest
    print("\nbest: ",DD[parsef(best)])
    f.write(str(size)+" "+str(tbest)+"\n")
    print("nb. flops: ",nflops, ", Gflops/s (best): ",flops/(10**9))
    print("-------")
    size*=2
    print(" ")
f.close()
